### DISCUSSION TITLE
是否依赖icudtl.dat

### DISCUSSION THREAD

**Discussion by xland** [Tencent/tgfx#384] [Q&A] [answered]
首先感谢作者开源如此强大的库。

据我所知skia在渲染大段、复杂文本时需要依赖icudtl.dat、icu.lib。
请问tgfx 在渲染大段、复杂文本时是否也需要依赖icudtl.dat、icu.lib？
或者说，tgfx不适用于渲染大段复杂文本的场景？

我也觉得Skia库太大了，再加上一个icudtl.dat(10M左右)，就更大了。
谢谢！

**Comment by domchen** [comment, marked as answer]
那个是文本塑型才用的，文字排版是另外一个框架的事，tgfx和skia核心引擎都只绘制排版完的glyphs结果，并不会依赖这些库，我们自己之前用skia也没有这个依赖，你要是有这个依赖，大概率是编译参数用错了，引用了了skparagrah这类扩展库。

**Comment by xland** [comment]
@domchen 

感谢回复。

我遇到的问题是：
1、我使用Skia开发了一个简单的GUI框架，涉及到文本输入框input控件，我发现输入光标的位置不能简单的通过SkFont.measureText方法计算，因为这样算出来的结果不准，尤其是中英文混合的情况下，经常会出现输入光标覆盖在文字之上的情况。
2、skparagrah提供了很多挺有用的能力，比如设置段落文本宽度，对齐方式，段落样式等。
请问，tgfx是否会提供类似的能力，弥补skia的不足。
Thanks very much.


下图是skparagrah设置段落宽度的能力。
如果不依赖这个库，单靠skia来完成类似的功能，想来应该挺难的。
![大段文本](https://github.com/user-attachments/assets/8bb3c4e4-a38d-4e0e-add5-a6aee5ace5a2)

**Reply by domchen** [reply]
这个不属于渲染引擎的范畴，tgfx不会去做这个资已经比较成熟的领域。复杂文本排版有很多成熟框架，比如你可以试试minikin。Skia写的这个据我所知也没有用在什么实际产品里，似乎只是写出来测试用的，功能也不完善，安卓用的也是minikin。

**Reply by domchen** [reply]
你如果需求比较简单并不是特别复杂的文本排版，想要更加精简的实现，其实就是在tgfx上自己再整合harfbuzz这个库就行了。skparagrah背后也是包装了harfbuzz，可以对照着抽取你需要的实现。你甚至可以用系统提供的接口去完成文本相关的测量排版，比如mac的coreText这种接口，然后提取出来glyphs交给引擎绘制，这个反而包体更小了。

**Reply by xland** [reply]
多谢回复，
我再深入研究一下~
